### PR TITLE
ch4/ofi: remove MPIDI_OFI_av_insert wrapper

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

When this wrapper was initially introduced, it has a couple #ifdefs --
now removed -- and it doesn't have the current vni_local/vni_remote
complications. Now that the #ifdefs are removed, and the actual usage
have more variations that made this wrapper function no longer useful.

Not only that, as currently written, is incorrect. It works because
we only used it for the vni=0 case but not other cases.

This commit removes this wrapper. If there will be future need -- for
example, the av AV struct grows sufficent sophistication that warrants
some wrapper, we can re-introduce it with more correctly tuned
interface.

[skip warnings]
## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

File: src/mpid/ch4/netmod/ofi/ofi_capability_sets.h
Comment (wesbland) [inline src/mpid/ch4/netmod/ofi/ofi_capability_sets.h:487]: We do use this internally for scalable endpoints vs. regular endpoints. I'm dragging in @sagarth and/or @shawnccx to say whether or not this would be an issue.

Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_capability_sets.h:487]: This MACRO is not used. Quick search from mpich code today (intel's branch).

File: src/mpid/ch4/netmod/ofi/ofi_types.h
Comment (shawnccx) [inline src/mpid/ch4/netmod/ofi/ofi_types.h:141]: This deletion is the major conflict with our downstream code. We have lots of done work relying `ep_idx` to explore multi endpoints usage.

File: src/mpid/ch4/netmod/ofi/ofi_capability_sets.h
Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_capability_sets.h:487]: Yep. It was the `ep_idx` part that was the problem here, not this macro.

File: src/mpid/ch4/netmod/ofi/ofi_types.h
Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: I am trying to understand why we need it. In upstream code, this `ep_idx` is always `0`. Practically, when we need a sub-endpoint for a particular vni, the `ep_idx` is just the `vni` (or deterministic from `vni`), right? If you have more scalable end points beyond vni, then you'll need some code for the derivation anyway, since a single `ep_idx` here won't help you distinguish different sub-endpoint anyway, right?

Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: In our downstream code, RMA uses an additional scalable endpoint. Its transmit contexts (for RMA) may map to the same vni as pt2pt. That means `vni` is not `deterministic` enough to determine which `ep` in the backend.

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: Interesting. So you don't use dedicated `rma_sep` for RMA? In upstream main code, we use `rma_sep` and store the context id per-window (inside `win`) so there isn't need to store `ep_idx` elsewhere. In https://github.com/pmodels/mpich/pull/4770, I am expanding this  `rma_sep` to be per-vni. 

I am still a bit confused even in the case we use global `sep`. When you map `RMA` to the same vni as pt2pt, are you using the same transmission context or separate one for RMA? Either way, the `ep_idx` in the `av` only can point to the  same `context`, so you will need extra code to determine the actual `ep_idx` anyway, right?

Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: We use dedicated `rma_sep`, but contexts for RMA are bound to the same group of completion queues as `pt2pt`. And our code refers idx for completion queues same as `vni` idx.

In this case, let's say we have both `pt2pt_sep` and `rma_sep`, so:
ctx0 in `pt2pt_sep` is bound to cq 0, which is vni0
ctx0 in `rma_sep` is bound to cq 0, which is vni0.

I think your design is another alternative for supporting multi-contexts in a particular `rma_sep`, maybe in future our downstream code will switch to the same as main. (we also don't want keep too much difference in our own branch).

Last questions above, from the description I just said, `RMA` uses a separate transmission (when scalable endpoint is supported). 

Also, one point in our code is that, the same `av` is bound to `pt2pt_sep` and `rma_sep`, in that case, looking for an `addr` in `av` doesn't require `ep_idx`. I think you are right to delete `ep_idx` for `av_lookup`.

```
MPIDI_OFI_CALL_RETURN(fi_scalable_ep_bind
                          (MPIDI_OFI_global.rma_sep, &(MPIDI_OFI_global.ctx[0].av->fid), 0), ret);
```

Let me forward this information to our team, to see if I mis-undestood anything.

Comment (raffenet) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: Is there any resolution to this discussion? I see this PR is waiting on a review from me, but I think the downstream conflicts need to be understood first.

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: @sagarth ?

Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: I'm waiting on @wesbland to see if this change conflicts his downstream multi-nic mapping.

Comment (hzhou) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: Oh, sorry. I guess I should tag @shawnccx ?

Comment (wesbland) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: Yeah, this isn't as big a deal for the multi-nic stuff since that's already part of the address information. The deletion of `MPIDI_OFI_av_insert` has more impact there, but since it isn't used much, I can work around that. This is up to @shawnccx.

Comment (shawnccx) [reply src/mpid/ch4/netmod/ofi/ofi_types.h:141]: Then I think we can move forward.

**Review by shawnccx — APPROVED** [review]

**Review by wesbland — APPROVED** [review]

**Review by raffenet — APPROVED** [review]

**Pull request merged**
